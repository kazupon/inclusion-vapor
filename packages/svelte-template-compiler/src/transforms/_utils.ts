import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
