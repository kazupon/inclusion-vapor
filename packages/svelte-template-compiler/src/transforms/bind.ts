// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vBind.ts

import {
  ErrorCodes,
  NodeTypes,
  createCompilerError,
  createSimpleExpression
} from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import { isReservedProp, resolveExpression } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { TransformContext } from './context.ts'
import type { DirectiveTransform } from './types.ts'

// same-name shorthand - :arg is expanded to :arg="arg"
export function normalizeBindShorthand(
  arg: SimpleExpressionNode,
  context: TransformContext
): SimpleExpressionNode {
  if (arg.type !== NodeTypes.SIMPLE_EXPRESSION || !arg.isStatic) {
    // only simple expression is allowed for same-name shorthand
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_BIND_INVALID_SAME_NAME_ARGUMENT, arg.loc)
    )
    return createSimpleExpression('', true, arg.loc)
  }

  const propName = camelize(arg.content)
  const exp = createSimpleExpression(propName, false, arg.loc)
  exp.ast = null // eslint-disable-line unicorn/no-null
  return exp
}

export const transformVBind: DirectiveTransform = (dir, _node, context) => {
  const { loc, modifiers } = dir
  let { exp } = dir
  let arg = dir.arg!
  const modifiersString = new Set(modifiers.map(s => s.content))

  if (!exp) {
    exp = normalizeBindShorthand(arg, context)
  }
  if (!exp.content.trim()) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
    exp = createSimpleExpression('', true, loc)
  }

  exp = resolveExpression(exp)
  arg = resolveExpression(arg)

  if (arg.isStatic && isReservedProp(arg.content)) {
    return
  }

  let camel = false
  if (modifiersString.has('camel')) {
    if (arg.isStatic) {
      arg = extend({}, arg, { content: camelize(arg.content) })
    } else {
      camel = true
    }
  }

  return {
    key: arg,
    value: exp,
    loc,
    runtimeCamelize: camel,
    modifier: modifiersString.has('prop') ? '.' : modifiersString.has('attr') ? '^' : undefined
  }
}
