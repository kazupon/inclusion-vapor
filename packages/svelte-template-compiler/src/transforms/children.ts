// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
