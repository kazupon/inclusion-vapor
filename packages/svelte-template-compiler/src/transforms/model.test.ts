import { NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { describe, expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformVBind } from './bind.ts'
import { transformChildren } from './children.ts'
import { transformComment } from './comment.ts'
import { transformElement } from './element.ts'
import { transformVModel } from './model.ts'
import { transformVOn } from './on.ts'
import { transformText } from './text.ts'

const compileWithVModel = makeCompile({
  prefixIdentifiers: false,
  nodeTransforms: [transformElement, transformChildren, transformText, transformComment],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
    model: transformVModel
  }
})

describe('bind:property', () => {
  describe('simple binding', () => {
    test('input', () => {
      const source1 = `<input bind:value={text} />`
      const source2 = `<input v-model="text" />`
      const { code, vaporHelpers, ir, helpers } = compileWithVModel(source1)
      const expectedResult = vaporCompile(source2)

      expect(code).toMatchSnapshot('received')
      expect(expectedResult.code).toMatchSnapshot('expected')

      expect(code).contains(`_withDirectives(n0, [[_vModelText, () => text]])`)
      expect(code).contains(`"update:modelValue", () => $event => (text = $event))`)

      expect(vaporHelpers).toContain('vModelText')
      expect(helpers.size).toBe(0)
      expect(ir.template).toEqual(['<input>'])
      expect(ir.block.operation).toMatchObject([
        {
          type: IRNodeTypes.SET_MODEL_VALUE,
          element: 0,
          isComponent: false,
          key: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'modelValue',
            isStatic: true
          },
          value: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'text',
            isStatic: false
          }
        },
        {
          type: IRNodeTypes.WITH_DIRECTIVE,
          name: 'vModelText',
          element: 0,
          builtin: true,
          dir: {
            arg: undefined,
            exp: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: 'text',
              isStatic: false
            }
          }
        }
      ])
    })

    test('textarea', () => {
      const source1 = `<textarea bind:value={text} />`
      const source2 = `<textarea v-model="text" />`
      const { code, ir, vaporHelpers } = compileWithVModel(source1)
      const expectedResult = vaporCompile(source2)

      expect(code).toMatchSnapshot('received')
      expect(expectedResult.code).toMatchSnapshot('expected')

      expect(code).contains(`_withDirectives(n0, [[_vModelText, () => text]])`)
      expect(code).contains(`"update:modelValue", () => $event => (text = $event))`)

      expect(vaporHelpers).toContain('vModelText')
      expect(ir.template).toEqual(['<textarea></textarea>'])
    })

    test('checkbox', () => {
      const source1 = `<input type="checkbox" bind:checked={yes} />`
      const source2 = `<input type="checkbox" v-model="yes" />`
      const { code, ir, vaporHelpers } = compileWithVModel(source1)
      const expectedResult = vaporCompile(source2)

      expect(code).toMatchSnapshot('received')
      expect(expectedResult.code).toMatchSnapshot('expected')

      expect(code).contains(`_withDirectives(n0, [[_vModelCheckbox, () => yes]])`)
      expect(code).contains(`"update:modelValue", () => $event => (yes = $event))`)

      expect(vaporHelpers).toContain('vModelCheckbox')
      expect(ir.template).toEqual(['<input type="checkbox">'])
    })
  })

  test('omit property', () => {
    const source = `<input bind:value />`
    const { code } = compileWithVModel(source)

    expect(code).toMatchSnapshot('received')

    expect(code).contains(`_withDirectives(n0, [[_vModelText, () => value]])`)
    expect(code).contains(`"update:modelValue", () => $event => (value = $event))`)
  })

  test('input number', () => {
    const source1 = `<input type="number" bind:value={num} />`
    const source2 = `<input type="number" v-model.number="num" />`
    const { code } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(
      `_withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])`
    )
    expect(code).contains(`"update:modelValue", () => $event => (num = $event))`)
  })

  test('input range', () => {
    const source1 = `<input type="range" bind:value={num} />`
    const source2 = `<input type="range" v-model.number="num" />`
    const { code } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(
      `_withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])`
    )
    expect(code).contains(`"update:modelValue", () => $event => (num = $event))`)
  })

  test.todo('input files', () => {
    const source1 = `<input accept="image/png, image/jpeg" bind:files id="avatar" name="avatar" type="file" />`
    expect(source1).toBe('todo')
  })

  test.todo('with on', () => {
    const source1 = `<input
	on:input={() => console.log('Old value:', value)}
	bind:value
	on:input={() => console.log('New value:', value)}
/>`
    expect(source1).toBe('todo')
  })

  test.todo('component', () => {
    const source1 = `<Keypad bind:value={pin} />`
    expect(source1).toBe('todo')
  })
})

describe('Binding <select> value', () => {
  test('basic', () => {
    const source1 = `<select bind:value={selected}>
  <option value={a}>a</option>
  <option value={b}>b</option>
  <option value={c}>c</option>
</select>`
    const source2 = `<select v-model="selected">
  <option :value="a">a</option>
  <option :value="b">b</option>
  <option :value="c">c</option>
</select>`
    const { code, ir, vaporHelpers } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(`_withDirectives(n3, [[_vModelSelect, () => selected]])`)
    expect(code).contains(`"update:modelValue", () => $event => (selected = $event))`)
    expect(code).contains(`_renderEffect(() => _setDynamicProp(n0, "value", a))`)
    expect(code).contains(`_renderEffect(() => _setDynamicProp(n1, "value", b))`)
    expect(code).contains(`_renderEffect(() => _setDynamicProp(n2, "value", c))`)

    expect(vaporHelpers).toContain('vModelSelect')
    expect(ir.template).toEqual([
      '<select><option>a</option> <option>b</option> <option>c</option></select>'
    ])
  })

  test('multiple', () => {
    const source1 = `<select multiple bind:value={fillings}>
	<option value="Rice">Rice</option>
	<option value="Beans">Beans</option>
	<option value="Cheese">Cheese</option>
	<option value="Guac (extra)">Guac (extra)</option>
</select>`
    const source2 = `<select multiple v-model="fillings">
	<option value="Rice">Rice</option>
	<option value="Beans">Beans</option>
	<option value="Cheese">Cheese</option>
	<option value="Guac (extra)">Guac (extra)</option>
</select>`
    const { code } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(`_withDirectives(n0, [[_vModelSelect, () => fillings]])`)
    expect(code).contains(`"update:modelValue", () => $event => (fillings = $event))`)
  })

  test('omit value', () => {
    const source1 = `<select multiple bind:value={fillings}>
	<option>Rice</option>
	<option>Beans</option>
	<option>Cheese</option>
	<option>Guac (extra)</option>
</select>`
    const { code } = compileWithVModel(source1)

    expect(code).toMatchSnapshot('received')

    expect(code).contains(`_withDirectives(n0, [[_vModelSelect, () => fillings]])`)
    expect(code).contains(`"update:modelValue", () => $event => (fillings = $event))`)
  })
})

describe.todo('contenteditable', () => {
  test('innerHTML', () => {
    const source1 = `<div contenteditable="true" bind:innerHTML={html} />`
    expect(source1).toBe('todo')
  })

  test('innerText', () => {
    const source1 = `<div contenteditable="true" bind:innerText />`
    expect(source1).toBe('todo')
  })

  test('textContent', () => {
    const source1 = `<div contenteditable="true" bind:textContent={text} />`
    expect(source1).toBe('todo')
  })
})

test.todo('details', () => {
  const source1 = `<details bind:open={isOpen}>
	<summary>Details</summary>
	<p>Something small enough to escape casual notice.</p>
</details>`
  expect(source1).toBe('todo')
})

describe.todo('Media element bindings', () => {
  test('video', () => {
    const source1 = `<video src={clip}
	bind:duration
	bind:buffered
	bind:played
	bind:seekable
	bind:seeking
	bind:ended
	bind:readyState
	bind:currentTime
	bind:playbackRate
	bind:paused
	bind:volume
	bind:muted
	bind:videoWidth
	bind:videoHeight
/>`
    expect(source1).toBe('todo')
  })

  test.todo('audio', () => {
    // TODO:
  })
})

test.todo('Image element bindings', () => {
  const source1 = `<img
	bind:naturalWidth
	bind:naturalHeight
></img>`
  expect(source1).toBe('todo')
})

test.todo('Block-level element bindings', () => {
  const source1 = `<div bind:offsetWidth={width} bind:offsetHeight={height}>
	<Chart {width} {height} />
</div>`
  expect(source1).toBe('todo')
})

describe('bind:group', () => {
  test('radio', () => {
    const source1 = `<input type="radio" bind:group={tortilla} value="Plain" />
<input type="radio" bind:group={tortilla} value="Whole wheat" />
<input type="radio" bind:group={tortilla} value="Spinach" />`
    const source2 = `<input type="radio" v-model="tortilla" value="Plain" />
<input type="radio" v-model="tortilla" value="Whole wheat" />
<input type="radio" v-model="tortilla" value="Spinach" />`
    const { code } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(`[[_vModelRadio, () => tortilla]]`)
    expect(code).contains(`"update:modelValue", () => $event => (tortilla = $event))`)
  })

  test('checkbox', () => {
    const source1 = `<input type="checkbox" bind:group={tortilla} value="Plain" />
<input type="checkbox" bind:group={tortilla} value="Whole wheat" />
<input type="checkbox" bind:group={tortilla} value="Spinach" />`
    const source2 = `<input type="checkbox" v-model="tortilla" value="Plain" />
<input type="checkbox" v-model="tortilla" value="Whole wheat" />
<input type="checkbox" v-model="tortilla" value="Spinach" />`
    const { code } = compileWithVModel(source1)
    const expectedResult = vaporCompile(source2)

    expect(code).toMatchSnapshot('received')
    expect(expectedResult.code).toMatchSnapshot('expected')

    expect(code).contains(`[[_vModelCheckbox, () => tortilla]]`)
    expect(code).contains(`"update:modelValue", () => $event => (tortilla = $event))`)
  })
})
