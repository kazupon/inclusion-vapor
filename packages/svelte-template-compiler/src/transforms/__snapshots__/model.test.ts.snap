// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Binding <select> value > basic > svelte 1`] = `
"import { vModelSelect as _vModelSelect, withDirectives as _withDirectives, next as _next, delegate as _delegate, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<select><option>a</option> <option>b</option> <option>c</option></select>")

export function render(_ctx) {
  const n3 = t0()
  _withDirectives(n3, [[_vModelSelect, () => selected]])
  const n0 = n3.firstChild
  const n1 = _next(n0, 2)
  const n2 = _next(n1, 2)
  _delegate(n3, "update:modelValue", () => $event => (selected = $event))
  _renderEffect(() => _setDynamicProp(n0, "value", a))
  _renderEffect(() => _setDynamicProp(n1, "value", b))
  _renderEffect(() => _setDynamicProp(n2, "value", c))
  return n3
}"
`;

exports[`Binding <select> value > basic > svelte 2`] = `
"import { vModelSelect as _vModelSelect, withDirectives as _withDirectives, delegate as _delegate, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<select><option>a</option><option>b</option><option>c</option></select>")

export function render(_ctx) {
  const n3 = t0()
  _withDirectives(n3, [[_vModelSelect, () => selected]])
  const n0 = n3.firstChild
  const n1 = n0.nextSibling
  const n2 = n1.nextSibling
  _delegate(n3, "update:modelValue", () => $event => (selected = $event))
  _renderEffect(() => _setDynamicProp(n0, "value", a))
  _renderEffect(() => _setDynamicProp(n1, "value", b))
  _renderEffect(() => _setDynamicProp(n2, "value", c))
  return n3
}"
`;

exports[`Binding <select> value > multiple > svelte 1`] = `
"import { vModelSelect as _vModelSelect, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<select multiple><option value=\\"Rice\\">Rice</option> <option value=\\"Beans\\">Beans</option> <option value=\\"Cheese\\">Cheese</option> <option value=\\"Guac (extra)\\">Guac (extra)</option></select>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelSelect, () => fillings]])
  _delegate(n0, "update:modelValue", () => $event => (fillings = $event))
  return n0
}"
`;

exports[`Binding <select> value > multiple > svelte 2`] = `
"import { vModelSelect as _vModelSelect, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<select multiple><option value=\\"Rice\\">Rice</option><option value=\\"Beans\\">Beans</option><option value=\\"Cheese\\">Cheese</option><option value=\\"Guac (extra)\\">Guac (extra)</option></select>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelSelect, () => fillings]])
  _delegate(n0, "update:modelValue", () => $event => (fillings = $event))
  return n0
}"
`;

exports[`Binding <select> value > omit value > svelte 1`] = `
"import { vModelSelect as _vModelSelect, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<select multiple><option>Rice</option> <option>Beans</option> <option>Cheese</option> <option>Guac (extra)</option></select>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelSelect, () => fillings]])
  _delegate(n0, "update:modelValue", () => $event => (fillings = $event))
  return n0
}"
`;

exports[`Component binding > basic > svelte 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Keypad = _resolveComponent("Keypad")
  const n0 = _createComponent(_component_Keypad, [
    { modelValue: () => (pin),
    "onUpdate:modelValue": () => $event => (pin = $event) }
  ], null, true)
  return n0
}"
`;

exports[`Component binding > basic > svelte 2`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Keypad = _resolveComponent("Keypad")
  const n0 = _createComponent(_component_Keypad, [
    { modelValue: () => (pin),
    "onUpdate:modelValue": () => $event => (pin = $event) }
  ], null, true)
  return n0
}"
`;

exports[`Component binding > multiple bindings > svelte 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_MyComp = _resolveComponent("MyComp")
  const n0 = _createComponent(_component_MyComp, [
    {
      foo: () => (buz),
      "onUpdate:foo": () => $event => (buz = $event), 
      bar: () => (bar),
      "onUpdate:bar": () => $event => (bar = $event)
    }
  ], null, true)
  return n0
}"
`;

exports[`Component binding > multiple bindings > svelte 2`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_MyComp = _resolveComponent("MyComp")
  const n0 = _createComponent(_component_MyComp, [
    {
      foo: () => (buz),
      "onUpdate:foo": () => $event => (buz = $event), 
      bar: () => (bar),
      "onUpdate:bar": () => $event => (bar = $event)
    }
  ], null, true)
  return n0
}"
`;

exports[`bind:group > checkbox > svelte 1`] = `
"import { vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\" value=\\"Plain\\">")
const t1 = _template(" ")
const t2 = _template("<input type=\\"checkbox\\" value=\\"Whole wheat\\">")
const t3 = _template("<input type=\\"checkbox\\" value=\\"Spinach\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelCheckbox, () => tortilla]])
  const n1 = t1()
  const n2 = t2()
  _withDirectives(n2, [[_vModelCheckbox, () => tortilla]])
  const n3 = t1()
  const n4 = t3()
  _withDirectives(n4, [[_vModelCheckbox, () => tortilla]])
  _delegate(n0, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n2, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n4, "update:modelValue", () => $event => (tortilla = $event))
  return [n0, n1, n2, n3, n4]
}"
`;

exports[`bind:group > checkbox > svelte 2`] = `
"import { vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\" value=\\"Plain\\">")
const t1 = _template("<input type=\\"checkbox\\" value=\\"Whole wheat\\">")
const t2 = _template("<input type=\\"checkbox\\" value=\\"Spinach\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelCheckbox, () => tortilla]])
  const n1 = t1()
  _withDirectives(n1, [[_vModelCheckbox, () => tortilla]])
  const n2 = t2()
  _withDirectives(n2, [[_vModelCheckbox, () => tortilla]])
  _delegate(n0, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n1, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n2, "update:modelValue", () => $event => (tortilla = $event))
  return [n0, n1, n2]
}"
`;

exports[`bind:group > radio > svelte 1`] = `
"import { vModelRadio as _vModelRadio, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"radio\\" value=\\"Plain\\">")
const t1 = _template(" ")
const t2 = _template("<input type=\\"radio\\" value=\\"Whole wheat\\">")
const t3 = _template("<input type=\\"radio\\" value=\\"Spinach\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelRadio, () => tortilla]])
  const n1 = t1()
  const n2 = t2()
  _withDirectives(n2, [[_vModelRadio, () => tortilla]])
  const n3 = t1()
  const n4 = t3()
  _withDirectives(n4, [[_vModelRadio, () => tortilla]])
  _delegate(n0, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n2, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n4, "update:modelValue", () => $event => (tortilla = $event))
  return [n0, n1, n2, n3, n4]
}"
`;

exports[`bind:group > radio > svelte 2`] = `
"import { vModelRadio as _vModelRadio, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"radio\\" value=\\"Plain\\">")
const t1 = _template("<input type=\\"radio\\" value=\\"Whole wheat\\">")
const t2 = _template("<input type=\\"radio\\" value=\\"Spinach\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelRadio, () => tortilla]])
  const n1 = t1()
  _withDirectives(n1, [[_vModelRadio, () => tortilla]])
  const n2 = t2()
  _withDirectives(n2, [[_vModelRadio, () => tortilla]])
  _delegate(n0, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n1, "update:modelValue", () => $event => (tortilla = $event))
  _delegate(n2, "update:modelValue", () => $event => (tortilla = $event))
  return [n0, n1, n2]
}"
`;

exports[`bind:property > input number > svelte 1`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"number\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])
  _delegate(n0, "update:modelValue", () => $event => (num = $event))
  return n0
}"
`;

exports[`bind:property > input number > svelte 2`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"number\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])
  _delegate(n0, "update:modelValue", () => $event => (num = $event))
  return n0
}"
`;

exports[`bind:property > input range > svelte 1`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"range\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])
  _delegate(n0, "update:modelValue", () => $event => (num = $event))
  return n0
}"
`;

exports[`bind:property > input range > svelte 2`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"range\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => num, void 0, { number: true }]])
  _delegate(n0, "update:modelValue", () => $event => (num = $event))
  return n0
}"
`;

exports[`bind:property > omit property > svelte 1`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => value]])
  _delegate(n0, "update:modelValue", () => $event => (value = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > checkbox > svelte 1`] = `
"import { vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelCheckbox, () => yes]])
  _delegate(n0, "update:modelValue", () => $event => (yes = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > checkbox > svelte 2`] = `
"import { vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\">")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelCheckbox, () => yes]])
  _delegate(n0, "update:modelValue", () => $event => (yes = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > input > svelte 1`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => text]])
  _delegate(n0, "update:modelValue", () => $event => (text = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > input > svelte 2`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => text]])
  _delegate(n0, "update:modelValue", () => $event => (text = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > textarea > svelte 1`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<textarea></textarea>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => text]])
  _delegate(n0, "update:modelValue", () => $event => (text = $event))
  return n0
}"
`;

exports[`bind:property > simple binding > textarea > svelte 2`] = `
"import { vModelText as _vModelText, withDirectives as _withDirectives, delegate as _delegate, template as _template } from 'vue/vapor';
const t0 = _template("<textarea></textarea>")

export function render(_ctx) {
  const n0 = t0()
  _withDirectives(n0, [[_vModelText, () => text]])
  _delegate(n0, "update:modelValue", () => $event => (text = $event))
  return n0
}"
`;
